blog/_posts/2021-10-30-debuging-root-scripts-in-eclipse.md

```diff
@@ -0,0 +1,227 @@
+* Increase Eclipse initial and maximum heap limits. The screenshot below demonstrates the Eclipse heap use while indexing a ROOT and Geant4-based project. Memory use fluctuates between 1 and 2 GB.
+
+<center>
+{% include figure_image
+   img="/assets/images/eclipse-heap-use.png"
+   style="width: 100%"
+   caption="Eclipse IDE memory use during the source code indexing operation. Obtained with VisualVm application."
+%}
+</center>
+
+Therefore, we will limit the heap to 2 GB. Locate your `eclipse.ini` configuration file. On my Fedora machine, it is `/usr/lib/eclipse/eclipse.ini`. Sometimes `eclipse.ini` is located inside your Eclipse installation folder. With text editor set minimum and maximum heap limits respectively: `-Xms512m` and `-Xmx2048`. Alternatively one can use `sed` with `-r` REGEXP parameter (double-check your `eclipse.ini` location below):
+
+```
+sudo cp /usr/lib/eclipse/eclipse.ini /usr/lib/eclipse/eclipse.ini.bak
+sudo sed -i -r "s;Xms[0-9]*m;Xms512m;" /usr/lib/eclipse/eclipse.ini
+sudo sed -i -r "s;Xmx[0-9]*m;Xmx2048m;" /usr/lib/eclipse/eclipse.ini  
+```
+
+* Increase Eclipse indexer cache limits. In Eclipse menu: Window > Preferences > C/C++ > Indexer set the "Cache limits" to 75% and 2048 MB.
+* Prevent workspace refreshes. In Window > Preferences > General > Workspace, uncheck "Refresh on access". Eclipse may randomly start refreshing the workspace upon the first run of a new Run/Debug configuration. Workspace refresh triggers the restart of an indexer. Additionally, for a CMake-managed project (not CDT managed) folder-specific Refresh Policy settings are unavailable. Our CMake project will refresh all the sources including ROOT sources (if attached). This is an expensive operation that usually takes quite some time. We want to prevent this from happening.
+
+Last but not least, it is useful to turn on the indication of the current heap size used by Eclipse. This can be done in Window > Preferences > General > Show heap status. The heap size will be displayed in the status bar. If heap usage maxes out, increase it accordingly and restart Eclipse.
+
+## Generating Eclipse Project
+
+Technically it is possible to attach a debugger directly to the ROOT Cling interpreter `root.exe`. In my opinion, this approach is **rather confusing**, especially for a newcomer:
+
+* Debug entry point will be outside of your program scope, namely in `rmain.cxx`.
+* Your script C++ file is never compiled. Program flow is passed to the interpreter that reads and processes your file line-by-line. Therefore, **breakpoints set in your actual source C++ file will never fire up**.
+
+A more straightforward solution would be **turning our ROOT script into a standalone ROOT-based C++ program** with `main()` function. This will ensure the correct entry point and provide an **intuitive debugging flow**.
```

Comment:
```suggestion
A alternative and robust solution is to **turn our ROOT script into a standalone ROOT-based C++ program** with `main()` function. This will ensure the correct entry point and provide an **intuitive debugging flow**.
```